src/decoding.rs

```diff
@@ -286,3 +286,140 @@ pub fn decode_header(token: &str) -> Result<Header> {
     let (_, header) = expect_two!(message.rsplitn(2, '.'));
     Header::from_encoded(header)
 }
+
+/// A builder style JWT decoder
+///
+/// # Examples
+///
+/// ```
+/// use jsonwebtoken::{JwtDecoder, HmacSecret};
+/// use serde::{Serialize, Deserialize};
+///
+/// #[derive(Debug, Serialize, Deserialize)]
+/// struct Claims {
+///    sub: String,
+///    company: String,
+///    exp: usize,
+/// }
+///
+/// let token = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJiQGIuY29tIiwiY29tcGFueSI6IkFDTUUiLCJleHAiOjI1MzI1MjQ4OTF9.9r56oF7ZliOBlOAyiOFperTGxBtPykRQiWNFxhDCW98";
+///
+/// let hmac_secret = HmacSecret::from_secret(b"secret");
+///
+/// let claims = JwtDecoder::hs_256(hmac_secret)
+///     .unwrap()
+///     .decode::<Claims>(&token)
+///     .unwrap();
+/// ```
+pub struct JwtDecoder {
+    verifying_provider: Box<dyn JwtVerifier>,
+    validation: Validation,
```

Comment:
Should it be `Option<Validator>`? The default Validation is probably not useful for a lot of cases so might as well not have one

Reply:
I am currently reusing a lot of the current validation logic that makes use of fields in the `Validation` struct. So in the initializer for the `JwtDecoder` it is [creating](https://github.com/sidrubs/jsonwebtoken/blob/decoupled-crypto-backends/src/decoding.rs#L327) a "default" `Validation` struct based on the algorithm of the `verifying_provider`. This is then used for the validation logic. If we were to use an `Option<Validator>` we would still need to provide default values to use somewhere in the code.

If someone wants to provide custom validation logic they can call the `with_validation` builder style method to replace the "default" validation that is generated in the initializer.

Reply:
If one wanted to clean things up one could flatten the `Validation` fields and methods into the `JwtDecoder`, but because we want to keep the original `decode` function working the `Validation` code needs to be there, so might as well make use of it.

Reply:
We should have the user pass it explicitly to the constructor then, implicit default validation is not good.

Reply:
I think I should replace the builder with a function (see [here](https://github.com/Keats/jsonwebtoken/pull/410#issuecomment-2392466079)). This would remove these smaller complexities.